File: go/tracer_test.go
package gentrail

import (
	"context"
	"strings"
	"testing"

	"go.opentelemetry.io/otel/attribute"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/sdk/trace/tracetest"
)

func newTestTracer(t *testing.T) (*Tracer, *tracetest.SpanRecorder) {
	t.Helper()
	recorder := tracetest.NewSpanRecorder()
	provider := sdktrace.NewTracerProvider(sdktrace.WithSpanProcessor(recorder))
	return &Tracer{
		tracer:   provider.Tracer("aigentrail.test"),
		provider: provider,
	}, recorder
}

func attrMap(attrs []attribute.KeyValue) map[string]attribute.Value {
	m := make(map[string]attribute.Value, len(attrs))
	for _, kv := range attrs {
		m[string(kv.Key)] = kv.Value
	}
	return m
}

func TestStartInvocationSetsAllAttributes(t *testing.T) {
	tracer, recorder := newTestTracer(t)
	_, inv := tracer.StartInvocation(context.Background(), InvocationParams{
		AgentID:     "agent-1",
		AgentName:   "Classifier",
		JournalID:   "journal-xyz",
		UserMessage: "classify these tools",
	})
	inv.End(InvocationEndParams{
		Response:      "done",
		TotalTokens:   42,
		ToolCount:     3,
		IntegrityHash: "abc123",
	})

	spans := recorder.Ended()
	if len(spans) != 1 {
		t.Fatalf("want 1 span, got %d", len(spans))
	}
	span := spans[0]
	if span.Name() != "governance.invocation" {
		t.Errorf("name = %q, want governance.invocation", span.Name())
	}
	attrs := attrMap(span.Attributes())
	checks := map[string]attribute.Value{
		"openinference.span.kind":           attribute.StringValue("AGENT"),
		"aigentrail.agent.id":               attribute.StringValue("agent-1"),
		"agent.name":                        attribute.StringValue("Classifier"),
		"aigentrail.journal.id":             attribute.StringValue("journal-xyz"),
		"session.id":                        attribute.StringValue("journal-xyz"),
		"input.value":                       attribute.StringValue("classify these tools"),
		"source":                            attribute.StringValue("aigentrail-sdk"),
		"output.value":                      attribute.StringValue("done"),
		"aigentrail.invocation.status":      attribute.StringValue("ok"),
		"aigentrail.journal.integrity_hash": attribute.StringValue("abc123"),
		"llm.token_count.total":             attribute.Int64Value(42),
		"aigentrail.tool.count":             attribute.Int64Value(3),
	}
	for k, want := range checks {
		got, ok := attrs[k]
		if !ok {
			t.Errorf("missing attribute %q", k)
			continue
		}
		if got.Emit() != want.Emit() {
			t.Errorf("attr %q = %v, want %v", k, got.Emit(), want.Emit())
		}
	}
}

func TestEndInvocationDefaultsStatusToOK(t *testing.T) {
	tracer, recorder := newTestTracer(t)
	_, inv := tracer.StartInvocation(context.Background(), InvocationParams{AgentID: "a"})
	inv.End(InvocationEndParams{})

	attrs := attrMap(recorder.Ended()[0].Attributes())
	if attrs["aigentrail.invocation.status"].AsString() != "ok" {
		t.Errorf("status default = %q, want ok", attrs["aigentrail.invocation.status"].AsString())
	}
}

func TestEndInvocationCustomStatus(t *testing.T) {
	tracer, recorder := newTestTracer(t)
	_, inv := tracer.StartInvocation(context.Background(), InvocationParams{AgentID: "a"})
	inv.End(InvocationEndParams{Status: "blocked"})

	attrs := attrMap(recorder.Ended()[0].Attributes())
	if attrs["aigentrail.invocation.status"].AsString() != "blocked" {
		t.Errorf("status = %q, want blocked", attrs["aigentrail.invocation.status"].AsString())
	}
}

func TestRecordModelCallEmitsLLMSpan(t *testing.T) {
	tracer, recorder := newTestTracer(t)
	ctx, inv := tracer.StartInvocation(context.Background(), InvocationParams{AgentID: "a"})
	tracer.RecordModelCall(ctx, ModelCallParams{
		ModelID:      "mistral-small-latest",
		Prompt:       "hello",
		ResponseText: "hi",
		InputTokens:  10,
		OutputTokens: 5,
		LatencyMS:    123.4,
	})
	inv.End(InvocationEndParams{})

	spans := recorder.Ended()
	if len(spans) != 2 {
		t.Fatalf("want 2 spans, got %d", len(spans))
	}
	var model sdktrace.ReadOnlySpan
	for _, s := range spans {
		if s.Name() == "governance.model_call" {
			model = s
		}
	}
	if model == nil {
		t.Fatal("no governance.model_call span recorded")
	}
	attrs := attrMap(model.Attributes())
	if attrs["openinference.span.kind"].AsString() != "LLM" {
		t.Errorf("span.kind = %q, want LLM", attrs["openinference.span.kind"].AsString())
	}
	if attrs["llm.model_name"].AsString() != "mistral-small-latest" {
		t.Errorf("model_name = %q", attrs["llm.model_name"].AsString())
	}
	if attrs["llm.token_count.prompt"].AsInt64() != 10 {
		t.Errorf("prompt tokens = %d", attrs["llm.token_count.prompt"].AsInt64())
	}
	if attrs["llm.token_count.completion"].AsInt64() != 5 {
		t.Errorf("completion tokens = %d", attrs["llm.token_count.completion"].AsInt64())
	}
	if attrs["aigentrail.latency_ms"].AsFloat64() != 123.4 {
		t.Errorf("latency = %v", attrs["aigentrail.latency_ms"].AsFloat64())
	}
}

func TestRecordModelCallSkipsZeroLatency(t *testing.T) {
	tracer, recorder := newTestTracer(t)
	ctx, inv := tracer.StartInvocation(context.Background(), InvocationParams{AgentID: "a"})
	tracer.RecordModelCall(ctx, ModelCallParams{ModelID: "m"})
	inv.End(InvocationEndParams{})

	for _, s := range recorder.Ended() {
		if s.Name() != "governance.model_call" {
			continue
		}
		for _, kv := range s.Attributes() {
			if string(kv.Key) == "aigentrail.latency_ms" {
				t.Errorf("latency attr present when LatencyMS==0")
			}
		}
	}
}

func TestRecordToolCallNamesSpanAfterTool(t *testing.T) {
	tracer, recorder := newTestTracer(t)
	ctx, inv := tracer.StartInvocation(context.Background(), InvocationParams{AgentID: "a"})
	tracer.RecordToolCall(ctx, ToolCallParams{
		AgentID:    "a",
		AgentName:  "Agent",
		Name:       "search_web",
		Args:       `{"q":"hi"}`,
		Result:     "ok",
		DurationMS: 7.5,
	})
	inv.End(InvocationEndParams{})

	var tool sdktrace.ReadOnlySpan
	for _, s := range recorder.Ended() {
		if s.Name() == "search_web" {
			tool = s
		}
	}
	if tool == nil {
		t.Fatal("no tool-named span recorded")
	}
	attrs := attrMap(tool.Attributes())
	if attrs["openinference.span.kind"].AsString() != "TOOL" {
		t.Errorf("span.kind = %q", attrs["openinference.span.kind"].AsString())
	}
	if attrs["tool.name"].AsString() != "search_web" {
		t.Errorf("tool.name = %q", attrs["tool.name"].AsString())
	}
	if attrs["aigentrail.agent.id"].AsString() != "a" {
		t.Errorf("agent.id = %q", attrs["aigentrail.agent.id"].AsString())
	}
	if attrs["aigentrail.latency_ms"].AsFloat64() != 7.5 {
		t.Errorf("latency = %v", attrs["aigentrail.latency_ms"].AsFloat64())
	}
}

func TestChildSpansNestUnderInvocation(t *testing.T) {
	tracer, recorder := newTestTracer(t)
	ctx, inv := tracer.StartInvocation(context.Background(), InvocationParams{AgentID: "a"})
	tracer.RecordModelCall(ctx, ModelCallParams{ModelID: "m"})
	tracer.RecordToolCall(ctx, ToolCallParams{Name: "t"})
	inv.End(InvocationEndParams{})

	var invSpanID string
	for _, s := range recorder.Ended() {
		if s.Name() == "governance.invocation" {
			invSpanID = s.SpanContext().SpanID().String()
		}
	}
	if invSpanID == "" {
		t.Fatal("invocation span not recorded")
	}
	for _, s := range recorder.Ended() {
		if s.Name() == "governance.invocation" {
			continue
		}
		if s.Parent().SpanID().String() != invSpanID {
			t.Errorf("span %q parent = %s, want %s", s.Name(), s.Parent().SpanID(), invSpanID)
		}
	}
}

func TestTruncateRunesRespectsCodePoints(t *testing.T) {
	long := strings.Repeat("a", 600)
	if got := truncateRunes(long, 500); len(got) != 500 {
		t.Errorf("truncated len = %d, want 500", len(got))
	}
	multibyte := strings.Repeat("世", 600)
	got := truncateRunes(multibyte, 500)
	if runes := []rune(got); len(runes) != 500 {
		t.Errorf("multibyte rune count = %d, want 500", len(runes))
	}
	if short := truncateRunes("hi", 500); short != "hi" {
		t.Errorf("short string mangled: %q", short)
	}
}

func TestStartInvocationTruncatesLongUserMessage(t *testing.T) {
	tracer, recorder := newTestTracer(t)
	long := strings.Repeat("x", maxInputValueRunes+100)
	_, inv := tracer.StartInvocation(context.Background(), InvocationParams{
		AgentID:     "a",
		UserMessage: long,
	})
	inv.End(InvocationEndParams{})

	attrs := attrMap(recorder.Ended()[0].Attributes())
	if got := attrs["input.value"].AsString(); len(got) != maxInputValueRunes {
		t.Errorf("input.value len = %d, want %d", len(got), maxInputValueRunes)
	}
}

func TestRecordModelCallTruncatesLongOutput(t *testing.T) {
	tracer, recorder := newTestTracer(t)
	long := strings.Repeat("y", maxModelOutputRunes+100)
	ctx, inv := tracer.StartInvocation(context.Background(), InvocationParams{AgentID: "a"})
	tracer.RecordModelCall(ctx, ModelCallParams{ModelID: "m", ResponseText: long})
	inv.End(InvocationEndParams{})

	for _, s := range recorder.Ended() {
		if s.Name() != "governance.model_call" {
			continue
		}
		attrs := attrMap(s.Attributes())
		if got := attrs["output.value"].AsString(); len(got) != maxModelOutputRunes {
			t.Errorf("output.value len = %d, want %d", len(got), maxModelOutputRunes)
		}
	}
}

func TestSplitEndpointVariants(t *testing.T) {
	cases := []struct {
		in       string
		wantHost string
		wantPath string
	}{
		{"https://otel.gentrail.ai:4318", "otel.gentrail.ai:4318", "/v1/traces"},
		{"http://localhost:4318/", "localhost:4318", "/v1/traces"},
		{"https://otel.gentrail.ai:4318/custom", "otel.gentrail.ai:4318", "/custom/v1/traces"},
		{"host:4318", "host:4318", "/v1/traces"},
		{"https://otel.gentrail.ai", "otel.gentrail.ai:443", "/v1/traces"},
		{"http://localhost", "localhost:80", "/v1/traces"},
	}
	for _, c := range cases {
		host, path, err := splitEndpoint(c.in)
		if err != nil {
			t.Errorf("%q: %v", c.in, err)
			continue
		}
		if host != c.wantHost || path != c.wantPath {
			t.Errorf("%q: host=%q path=%q, want host=%q path=%q", c.in, host, path, c.wantHost, c.wantPath)
		}
	}
}

func TestSplitEndpointRejectsEmpty(t *testing.T) {
	if _, _, err := splitEndpoint(""); err == nil {
		t.Error("want error for empty endpoint")
	}
}

func TestNilTracerMethodsAreNoop(t *testing.T) {
	var tracer *Tracer
	ctx, inv := tracer.StartInvocation(context.Background(), InvocationParams{AgentID: "a"})
	tracer.RecordModelCall(ctx, ModelCallParams{ModelID: "m"})
	tracer.RecordToolCall(ctx, ToolCallParams{Name: "t"})
	tracer.RecordLLMCall(ctx, LLMCallParams{ModelID: "m"})
	inv.End(InvocationEndParams{})
	if err := tracer.ForceFlush(context.Background()); err != nil {
		t.Errorf("nil ForceFlush err = %v", err)
	}
	if err := tracer.Shutdown(context.Background()); err != nil {
		t.Errorf("nil Shutdown err = %v", err)
	}
}

func TestRecordLLMCallEmitsInvocationAndModelSpans(t *testing.T) {
	tracer, recorder := newTestTracer(t)
	tracer.RecordLLMCall(context.Background(), LLMCallParams{
		AgentID:      "agent-1",
		AgentName:    "Classifier",
		ModelID:      "mistral-small-latest",
		Prompt:       "classify",
		Response:     "done",
		InputTokens:  10,
		OutputTokens: 5,
		LatencyMS:    50,
	})

	spans := recorder.Ended()
	if len(spans) != 2 {
		t.Fatalf("want 2 spans (invocation + model_call), got %d", len(spans))
	}
	var invocation, model sdktrace.ReadOnlySpan
	for _, s := range spans {
		switch s.Name() {
		case "governance.invocation":
			invocation = s
		case "governance.model_call":
			model = s
		}
	}
	if invocation == nil || model == nil {
		t.Fatalf("missing spans: invocation=%v model=%v", invocation != nil, model != nil)
	}
	if got := attrMap(invocation.Attributes())["llm.token_count.total"].AsInt64(); got != 15 {
		t.Errorf("invocation total tokens = %d, want 15", got)
	}
	if got := attrMap(model.Attributes())["llm.model_name"].AsString(); got != "mistral-small-latest" {
		t.Errorf("model_name = %q", got)
	}
	if model.Parent().SpanID() != invocation.SpanContext().SpanID() {
		t.Error("model_call should nest under the invocation")
	}
}

func TestStartInvocationGeneratesJournalIDWhenEmpty(t *testing.T) {
	tracer, recorder := newTestTracer(t)
	_, inv := tracer.StartInvocation(context.Background(), InvocationParams{AgentID: "a"})
	inv.End(InvocationEndParams{})

	attrs := attrMap(recorder.Ended()[0].Attributes())
	if attrs["session.id"].AsString() == "" {
		t.Error("session.id empty; JournalID should be auto-generated when not supplied")
	}
	if attrs["aigentrail.journal.id"].AsString() == "" {
		t.Error("journal.id empty; should be auto-generated when not supplied")
	}
}

func TestNewWithoutAPIKeyReturnsErr(t *testing.T) {
	t.Setenv("GENTRAIL_API_KEY", "")
	_, err := New(context.Background())
	if err != ErrMissingAPIKey {
		t.Errorf("err = %v, want ErrMissingAPIKey", err)
	}
}

File: js/gentrail-ai/test/tracer.test.ts
import assert from "node:assert/strict";
import test from "node:test";

import {
  BasicTracerProvider,
  InMemorySpanExporter,
  SimpleSpanProcessor,
  type ReadableSpan,
} from "@opentelemetry/sdk-trace-base";

import { createGovernanceTracer, GovernanceTracer } from "../src/index.js";
import { readSpecJson, serve, withEnv } from "./support.js";

interface SpanSpec {
  name: string;
  span_kind: string;
  start_attributes?: string[];
  end_attributes?: string[];
  attributes?: string[];
  optional_attributes?: string[];
}

interface SpansSpec {
  source: string;
  value_rune_limit: number;
  enforcement_decision_attribute: string;
  invocation: SpanSpec;
  model_call: SpanSpec;
  tool_call: SpanSpec;
}

const spansSpec = readSpecJson("spans.json") as SpansSpec;

function inMemoryTracer(redact = true): {
  tracer: GovernanceTracer;
  exporter: InMemorySpanExporter;
} {
  const exporter = new InMemorySpanExporter();
  const tracerProvider = new BasicTracerProvider({
    spanProcessors: [new SimpleSpanProcessor(exporter)],
  });
  return { tracer: new GovernanceTracer({ tracerProvider, redact }), exporter };
}

function spanNamed(spans: ReadableSpan[], name: string): ReadableSpan {
  const span = spans.find((candidate) => candidate.name === name);
  assert.ok(span !== undefined, `no span named ${name}`);
  return span;
}

function assertAttributesPresent(span: ReadableSpan, keys: readonly string[]): void {
  for (const key of keys) {
    assert.ok(key in span.attributes, `${span.name} is missing ${key}`);
  }
}

function assertChildOf(child: ReadableSpan, parent: ReadableSpan): void {
  assert.equal(child.parentSpanContext?.spanId, parent.spanContext().spanId);
  assert.equal(child.spanContext().traceId, parent.spanContext().traceId);
}

function recordFullInvocation(tracer: GovernanceTracer, enforcedDecision?: string): void {
  const invocation = tracer.startInvocation({
    agentId: "agent-billing",
    agentName: "Billing Agent",
    journalId: "inv-1",
    userMessage: "refund jane@acme.com",
  });
  tracer.recordModelCall(invocation, {
    modelId: "claude-sonnet",
    prompt: "refund jane@acme.com",
    responseText: "calling issue_refund",
    inputTokens: 120,
    outputTokens: 48,
    latencyMs: 812.25,
  });
  tracer.recordToolCall(invocation, {
    agentId: "agent-billing",
    agentName: "Billing Agent",
    name: "issue_refund",
    args: '{"ssn":"123-45-6789"}',
    result: "ok",
    durationMs: 41.5,
    enforcedDecision,
  });
  tracer.endInvocation(invocation, {
    response: "Refund issued.",
    totalTokens: 168,
    toolCount: 1,
    integrityHash: "abc",
  });
}

test("governance spans carry every attribute pinned by spans.json", () => {
  const { tracer, exporter } = inMemoryTracer();
  recordFullInvocation(tracer, "BLOCK");
  const spans = exporter.getFinishedSpans();
  const invocation = spanNamed(spans, spansSpec.invocation.name);
  const modelCall = spanNamed(spans, spansSpec.model_call.name);
  const toolCall = spanNamed(spans, "issue_refund");
  assertAttributesPresent(invocation, [
    ...(spansSpec.invocation.start_attributes ?? []),
    ...(spansSpec.invocation.end_attributes ?? []),
  ]);
  assertAttributesPresent(modelCall, [
    ...(spansSpec.model_call.attributes ?? []),
    ...(spansSpec.model_call.optional_attributes ?? []),
  ]);
  assertAttributesPresent(toolCall, [
    ...(spansSpec.tool_call.attributes ?? []),
    ...(spansSpec.tool_call.optional_attributes ?? []),
  ]);
  assert.equal(invocation.attributes["openinference.span.kind"], spansSpec.invocation.span_kind);
  assert.equal(modelCall.attributes["openinference.span.kind"], spansSpec.model_call.span_kind);
  assert.equal(toolCall.attributes["openinference.span.kind"], spansSpec.tool_call.span_kind);
  assert.equal(invocation.attributes.source, spansSpec.source);
  assert.equal(invocation.attributes["session.id"], "inv-1");
  assert.equal(invocation.attributes["aigentrail.invocation.status"], "ok");
  assert.equal(toolCall.attributes["tool.name"], "issue_refund");
  assert.equal(toolCall.attributes[spansSpec.enforcement_decision_attribute], "BLOCK");
  assert.equal(modelCall.attributes["aigentrail.latency_ms"], 812.25);
  assertChildOf(modelCall, invocation);
  assertChildOf(toolCall, invocation);
  assert.equal(invocation.parentSpanContext, undefined);
});

test("optional attributes are omitted when not supplied", () => {
  const { tracer, exporter } = inMemoryTracer();
  const invocation = tracer.startInvocation({
    agentId: "a",
    agentName: "A",
    journalId: "j",
    userMessage: "",
  });
  tracer.recordModelCall(invocation, {
    modelId: "m",
    prompt: "",
    responseText: "",
    inputTokens: 0,
    outputTokens: 0,
  });
  tracer.recordToolCall(invocation, {
    agentId: "a",
    agentName: "A",
    name: "t",
    args: "",
    result: "",
  });
  tracer.endInvocation(invocation, {
    response: "",
    totalTokens: 0,
    toolCount: 1,
    integrityHash: "",
    status: "error",
  });
  const spans = exporter.getFinishedSpans();
  assert.equal(
    "aigentrail.latency_ms" in spanNamed(spans, spansSpec.model_call.name).attributes,
    false,
  );
  assert.equal("aigentrail.latency_ms" in spanNamed(spans, "t").attributes, false);
  assert.equal(spansSpec.enforcement_decision_attribute in spanNamed(spans, "t").attributes, false);
  assert.equal(
    spanNamed(spans, spansSpec.invocation.name).attributes["aigentrail.invocation.status"],
    "error",
  );
});

test("values are redacted unless disabled", () => {
  const redacting = inMemoryTracer(true);
  recordFullInvocation(redacting.tracer);
  const redactedSpans = redacting.exporter.getFinishedSpans();
  assert.equal(
    spanNamed(redactedSpans, spansSpec.invocation.name).attributes["input.value"],
    "refund [EMAIL]",
  );
  assert.equal(
    spanNamed(redactedSpans, "issue_refund").attributes["input.value"],
    '{"ssn":"[SSN]"}',
  );
  const raw = inMemoryTracer(false);
  recordFullInvocation(raw.tracer);
  assert.equal(
    spanNamed(raw.exporter.getFinishedSpans(), spansSpec.invocation.name).attributes["input.value"],
    "refund jane@acme.com",
  );
});

test("values are truncated to the code point limit after redaction", () => {
  const { tracer, exporter } = inMemoryTracer();
  const limit = spansSpec.value_rune_limit;
  const astral = "\u{1F600}".repeat(limit + 10);
  const invocation = tracer.startInvocation({
    agentId: "a",
    agentName: "A",
    journalId: "j",
    userMessage: astral,
  });
  tracer.endInvocation(invocation, {
    response: "x".repeat(limit - 2) + " a@b.com",
    totalTokens: 0,
    toolCount: 0,
    integrityHash: "",
  });
  const span = spanNamed(exporter.getFinishedSpans(), spansSpec.invocation.name);
  const input = String(span.attributes["input.value"]);
  assert.equal([...input].length, limit);
  assert.equal(input, "\u{1F600}".repeat(limit));
  assert.equal(span.attributes["output.value"], "x".repeat(limit - 2) + " [");
});

test("recordLLMCall emits a closed invocation with one model call", () => {
  const { tracer, exporter } = inMemoryTracer();
  tracer.recordLLMCall({
    modelId: "gpt",
    prompt: "hi",
    responseText: "hello",
    inputTokens: 3,
    outputTokens: 4,
  });
  const spans = exporter.getFinishedSpans();
  assert.equal(spans.length, 2);
  const invocation = spanNamed(spans, spansSpec.invocation.name);
  assertChildOf(spanNamed(spans, spansSpec.model_call.name), invocation);
  assert.match(String(invocation.attributes["aigentrail.journal.id"]), /^[0-9a-f]{32}$/);
  assert.equal(invocation.attributes["llm.token_count.total"], 7);
  assert.equal(invocation.attributes["aigentrail.tool.count"], 0);
});

test("fetch records one LLM call per request with its status", async () => {
  const srv = await serve((req) =>
    req.url === "/v1/fail" ? { status: 503 } : { json: { ok: true } },
  );
  try {
    const { tracer, exporter } = inMemoryTracer();
    const tracedFetch = tracer.fetch();
    const ok = await tracedFetch(`${srv.base}/v1/messages?x=1`, { method: "post", body: "{}" });
    assert.equal(ok.status, 200);
    const failed = await tracedFetch(new Request(`${srv.base}/v1/fail`));
    assert.equal(failed.status, 503);
    await assert.rejects(tracedFetch("http://127.0.0.1:9/v1/down"));
    const modelCalls = exporter
      .getFinishedSpans()
      .filter((span) => span.name === spansSpec.model_call.name);
    const host = new URL(srv.base).host;
    assert.deepEqual(
      modelCalls.map((span) => [
        span.attributes["llm.model_name"],
        span.attributes["input.value"],
        span.attributes["output.value"],
      ]),
      [
        [host, "POST /v1/messages", "ok"],
        [host, "GET /v1/fail", "http_503"],
        ["127.0.0.1:9", "GET /v1/down", "error"],
      ],
    );
    for (const span of modelCalls) {
      assert.ok(Number.isInteger(span.attributes["aigentrail.latency_ms"]));
    }
    const statuses = exporter
      .getFinishedSpans()
      .filter((span) => span.name === spansSpec.invocation.name)
      .map((span) => span.attributes["aigentrail.invocation.status"]);
    assert.deepEqual(statuses, ["ok", "http_503", "error"]);
  } finally {
    await srv.close();
  }
});

test("createGovernanceTracer stays off without GENTRAIL_API_KEY and warns when half configured", async (t) => {
  const warn = t.mock.method(console, "warn", () => undefined);
  await withEnv(
    {
      GENTRAIL_API_KEY: undefined,
      AIGENTRAIL_API_KEY: undefined,
      OTEL_EXPORTER_OTLP_ENDPOINT: undefined,
    },
    () => {
      assert.equal(createGovernanceTracer(), null);
    },
  );
  assert.equal(warn.mock.callCount(), 0);
  await withEnv(
    {
      GENTRAIL_API_KEY: undefined,
      AIGENTRAIL_API_KEY: "old",
      OTEL_EXPORTER_OTLP_ENDPOINT: undefined,
    },
    () => {
      assert.equal(createGovernanceTracer(), null);
    },
  );
  await withEnv(
    {
      GENTRAIL_API_KEY: undefined,
      AIGENTRAIL_API_KEY: undefined,
      OTEL_EXPORTER_OTLP_ENDPOINT: "http://x",
    },
    () => {
      assert.equal(createGovernanceTracer(), null);
    },
  );
  assert.equal(warn.mock.callCount(), 2);
});

async function exportOnceFromEnv(
  env: Record<string, string | undefined>,
): Promise<{ url: string; auth: string | undefined }> {
  const srv = await serve(() => ({ json: {} }));
  try {
    return await withEnv({ ...env, OTEL_EXPORTER_OTLP_ENDPOINT: srv.base + "/" }, async () => {
      const tracer = createGovernanceTracer();
      assert.ok(tracer !== null);
      tracer.recordLLMCall({ modelId: "m", prompt: "p", responseText: "r" });
      await tracer.shutdown();
      assert.equal(srv.requests.length, 1);
      return { url: srv.requests[0].url, auth: srv.requests[0].auth };
    });
  } finally {
    await srv.close();
  }
}

test("createGovernanceTracer exports to the configured endpoint with Bearer auth", async () => {
  const seen = await exportOnceFromEnv({
    GENTRAIL_API_KEY: "sk-env",
    OTEL_EXPORTER_OTLP_HEADERS: undefined,
    OTEL_EXPORTER_OTLP_TRACES_HEADERS: undefined,
  });
  assert.deepEqual(seen, { url: "/v1/traces", auth: "Bearer sk-env" });
});

test("standard OTLP header variables replace the Bearer header", async () => {
  const seen = await exportOnceFromEnv({
    GENTRAIL_API_KEY: "sk-env",
    OTEL_EXPORTER_OTLP_HEADERS: "authorization=Custom abc",
    OTEL_EXPORTER_OTLP_TRACES_HEADERS: undefined,
  });
  assert.deepEqual(seen, { url: "/v1/traces", auth: "Custom abc" });
});
